From 3935b5aa5f3b2b3c7ca7930c9f5c67e54c291c4e Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Mon, 24 Jul 2023 17:04:43 +0800
Subject: [PATCH 21/53] [Huawei] mm/sharepool: Extract sp_k2u_populate()

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

This is used to build the pagetable for the task during k2u.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 63 ++++++++++++++++++++++++++++---------------------
 1 file changed, 36 insertions(+), 27 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index d951bb55f066..f409e7423eb5 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -2419,53 +2419,31 @@ static unsigned long __sp_remap_get_pfn(unsigned long kva)
 	return pfn;
 }
 
-/* when called by k2u to group, always make sure rw_lock of spg is down */
-static int sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *mm,
-			       unsigned long prot, struct sp_k2u_context *kc)
+static int sp_k2u_populate(struct mm_struct *mm, struct sp_area *spa)
 {
-	int ret = 0;
-	unsigned long ret_addr;
+	int ret;
 	struct vm_area_struct *vma;
-	unsigned long populate = 0;
-	unsigned long addr, buf, offset;
 	unsigned long kva = spa->kva;
-
-	down_write(&mm->mmap_lock);
-	if (unlikely(mm->core_state)) {
-		up_write(&mm->mmap_lock);
-		pr_err("k2u mmap: encountered coredump, abort\n");
-		return SP_SKIP_ERR;
-	}
-
-	ret_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot);
-	if (IS_ERR_VALUE(ret_addr)) {
-		up_write(&mm->mmap_lock);
-		pr_info("k2u mmap failed %lx\n", ret_addr);
-		return (int)ret_addr;
-	}
+	unsigned long addr, buf, offset;
 
 	/* This should not fail because we hold the mmap_lock sicne mmap */
 	vma = find_vma(mm, spa->va_start);
 	if (is_vm_hugetlb_page(vma)) {
 		ret = remap_vmalloc_hugepage_range(vma, (void *)kva, 0);
 		if (ret) {
-			do_munmap(mm, ret_addr, spa_size(spa), NULL);
-			up_write(&mm->mmap_lock);
 			pr_debug("remap vmalloc hugepage failed, ret %d, kva is %lx\n",
 				 ret, (unsigned long)kva);
 			return ret;
 		}
 		vma->vm_flags |= VM_IO | VM_PFNMAP | VM_DONTEXPAND | VM_DONTDUMP;
 	} else {
-		buf = ret_addr;
 		addr = kva;
 		offset = 0;
+		buf = spa->va_start;
 		do {
 			ret = remap_pfn_range(vma, buf, __sp_remap_get_pfn(addr), PAGE_SIZE,
 					__pgprot(vma->vm_page_prot.pgprot));
 			if (ret) {
-				do_munmap(mm, ret_addr, spa_size(spa), NULL);
-				up_write(&mm->mmap_lock);
 				pr_err("remap_pfn_range failed %d\n", ret);
 				return ret;
 			}
@@ -2475,9 +2453,40 @@ static int sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *mm,
 		} while (offset < spa_size(spa));
 	}
 
+	return 0;
+}
+
+/* when called by k2u to group, always make sure rw_lock of spg is down */
+static int sp_remap_kva_to_vma(struct sp_area *spa, struct mm_struct *mm,
+			       unsigned long prot, struct sp_k2u_context *kc)
+{
+	int ret = 0;
+	unsigned long ret_addr;
+	unsigned long populate = 0;
+
+	down_write(&mm->mmap_lock);
+	if (unlikely(mm->core_state)) {
+		up_write(&mm->mmap_lock);
+		pr_err("k2u mmap: encountered coredump, abort\n");
+		return SP_SKIP_ERR;
+	}
+
+	ret_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot);
+	if (IS_ERR_VALUE(ret_addr)) {
+		up_write(&mm->mmap_lock);
+		pr_info("k2u mmap failed %lx\n", ret_addr);
+		return (int)ret_addr;
+	}
+
+	ret = sp_k2u_populate(mm, spa);
+	if (ret) {
+		do_munmap(mm, ret_addr, spa_size(spa), NULL);
+		pr_info("k2u populate failed, %d\n", ret);
+	}
+
 	up_write(&mm->mmap_lock);
 
-	return 0;
+	return ret;
 }
 
 static bool vmalloc_area_set_flag(unsigned long kva, unsigned long flags)
-- 
2.17.1

